# Simple Exercises for Filtering and Aggregations

### Load SQL extension

In [48]:
%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [49]:
%env DATABASE_URL=postgresql://itversity_retail_user:itversity@localhost:5432/itversity_retail_db

env: DATABASE_URL=postgresql://itversity_retail_user:itversity@localhost:5432/itversity_retail_db


In [50]:
%%sql

SELECT count(*) FROM departments;
# SELECT count(*) FROM categories;
# SELECT count(*) FROM products;
# SELECT count(*) FROM orders;
# SELECT count(*) FROM order_items;
# SELECT count(*) FROM customers;

Running query in 'postgresql://itversity_retail_user:***@localhost:5432/itversity_retail_db'

1 rows affected.

count
6


## Exercises

### 1. Customer order count
Get order count per customer for the month of 2014 January.
* Tables - `orders` and `customers`
* Data should be sorted in descending order by count and ascending order by customer id.
* Output should contain `customer_id`, `customer_fname`, `customer_lname` and `customer_order_count`.

In [51]:
%%sql 

SELECT c.customer_id, customer_fname, customer_lname, 
    count(*) AS coustomer_order_count
FROM customers AS c
    LEFT OUTER JOIN orders AS o 
    ON c.customer_id = o.order_customer_id
WHERE to_char(o.order_date::timestamp, 'yyyy-MM') = '2014-01'
GROUP BY 1, 2, 3
ORDER BY count(*) DESC, c.customer_id ASC;

Running query in 'postgresql://itversity_retail_user:***@localhost:5432/itversity_retail_db'

4696 rows affected.

customer_id,customer_fname,customer_lname,coustomer_order_count
8622,Shirley,Smith,5
9676,Theresa,Smith,5
7,Melissa,Wilcox,4
222,Frank,Ruiz,4
2444,Kenneth,Smith,4
2485,Mary,Hernandez,4
2555,Mary,Long,4
3128,Karen,Turner,4
3199,Ashley,Hernandez,4
3610,Jordan,Smith,4


### 2. Dormant Customers
Get the customer details who have not placed any order for the month of 2014 January.

* Tables - `orders` and `customers`
* Output Columns - **All columns from customers as is**
* Data should be sorted in ascending order by `customer_id`
* Output should contain all the fields from `customers`
* Make sure to run below provided validation queries and validate the output.

In [52]:
%%sql 

SELECT c.*
FROM customers AS c
    LEFT OUTER JOIN orders AS o 
    ON c.customer_id = o.order_customer_id
    AND to_char(o.order_date::timestamp, 'yyyy-MM') = '2014-01'
WHERE o.order_customer_id IS NULL
ORDER BY 1 ASC;

Running query in 'postgresql://itversity_retail_user:***@localhost:5432/itversity_retail_db'

7739 rows affected.

customer_id,customer_fname,customer_lname,customer_email,customer_password,customer_street,customer_city,customer_state,customer_zipcode
1,Richard,Hernandez,XXXXXXXXX,XXXXXXXXX,6303 Heather Plaza,Brownsville,TX,78521
2,Mary,Barrett,XXXXXXXXX,XXXXXXXXX,9526 Noble Embers Ridge,Littleton,CO,80126
3,Ann,Smith,XXXXXXXXX,XXXXXXXXX,3422 Blue Pioneer Bend,Caguas,PR,00725
4,Mary,Jones,XXXXXXXXX,XXXXXXXXX,8324 Little Common,San Marcos,CA,92069
5,Robert,Hudson,XXXXXXXXX,XXXXXXXXX,10 Crystal River Mall,Caguas,PR,00725
6,Mary,Smith,XXXXXXXXX,XXXXXXXXX,3151 Sleepy Quail Promenade,Passaic,NJ,07055
9,Mary,Perez,XXXXXXXXX,XXXXXXXXX,3616 Quaking Street,Caguas,PR,00725
10,Melissa,Smith,XXXXXXXXX,XXXXXXXXX,8598 Harvest Beacon Plaza,Stafford,VA,22554
11,Mary,Huffman,XXXXXXXXX,XXXXXXXXX,3169 Stony Woods,Caguas,PR,00725
12,Christopher,Smith,XXXXXXXXX,XXXXXXXXX,5594 Jagged Embers By-pass,San Antonio,TX,78227


Another solution:

In [53]:
%%sql

SELECT *
FROM customers AS c
WHERE NOT EXISTS(
    SELECT o.order_customer_id
    FROM orders AS o 
    WHERE c.customer_id = o.order_customer_id
    AND to_char(o.order_date::timestamp, 'yyyy-MM') = '2014-01'
)
ORDER BY 1 ASC;

Running query in 'postgresql://itversity_retail_user:***@localhost:5432/itversity_retail_db'

7739 rows affected.

customer_id,customer_fname,customer_lname,customer_email,customer_password,customer_street,customer_city,customer_state,customer_zipcode
1,Richard,Hernandez,XXXXXXXXX,XXXXXXXXX,6303 Heather Plaza,Brownsville,TX,78521
2,Mary,Barrett,XXXXXXXXX,XXXXXXXXX,9526 Noble Embers Ridge,Littleton,CO,80126
3,Ann,Smith,XXXXXXXXX,XXXXXXXXX,3422 Blue Pioneer Bend,Caguas,PR,00725
4,Mary,Jones,XXXXXXXXX,XXXXXXXXX,8324 Little Common,San Marcos,CA,92069
5,Robert,Hudson,XXXXXXXXX,XXXXXXXXX,10 Crystal River Mall,Caguas,PR,00725
6,Mary,Smith,XXXXXXXXX,XXXXXXXXX,3151 Sleepy Quail Promenade,Passaic,NJ,07055
9,Mary,Perez,XXXXXXXXX,XXXXXXXXX,3616 Quaking Street,Caguas,PR,00725
10,Melissa,Smith,XXXXXXXXX,XXXXXXXXX,8598 Harvest Beacon Plaza,Stafford,VA,22554
11,Mary,Huffman,XXXXXXXXX,XXXXXXXXX,3169 Stony Woods,Caguas,PR,00725
12,Christopher,Smith,XXXXXXXXX,XXXXXXXXX,5594 Jagged Embers By-pass,San Antonio,TX,78227


Check answer:

In [54]:
%%sql 

SELECT count(DISTINCT order_customer_id)
FROM orders
WHERE to_char(order_date, 'yyyy-MM') = '2014-01';

Running query in 'postgresql://itversity_retail_user:***@localhost:5432/itversity_retail_db'

1 rows affected.

count
4696


In [55]:
%%sql 
SELECT count(*)
FROM customers;

Running query in 'postgresql://itversity_retail_user:***@localhost:5432/itversity_retail_db'

1 rows affected.

count
12435


Get the difference  
Get the count using solution query, both the difference and this count should match

Difference check: 12435 - 4696 = 7739  
My answer: 7739 rows affected.

### 3. Revenue Per Customer

Get the revenue generated by each customer for the month of 2014 January
* Tables - `orders`, `order_items` and `customers`
* Data should be sorted in descending order by revenue and then ascending order by `customer_id`
* Output should contain `customer_id`, `customer_fname`, `customer_lname`, `customer_revenue`.
* If there are no orders placed by customer, then the corresponding revenue for a given customer should be 0.
* Consider only `COMPLETE` and `CLOSED` orders

In [61]:
%%sql 

SELECT c.customer_id, c.customer_fname, c.customer_lname, 
    COALESCE(SUM(oi.order_item_subtotal), 0) AS customer_revenue
FROM customers AS c
    LEFT OUTER JOIN orders AS o 
        ON c.customer_id = o.order_customer_id
        AND to_char(o.order_date::timestamp, 'yyyy-MM') = '2014-01'
        AND o.order_status IN ('COMPLETE' , 'CLOSED')
    LEFT OUTER JOIN order_items AS oi 
        ON o.order_id = oi.order_item_order_id
GROUP BY 1, 2, 3
ORDER BY customer_revenue DESC, 1 ASC;

Running query in 'postgresql://itversity_retail_user:***@localhost:5432/itversity_retail_db'

12435 rows affected.

customer_id,customer_fname,customer_lname,customer_revenue
2555,Mary,Long,2954.629999999999
3465,Mary,Gardner,2929.7400000000002
3710,Ashley,Smith,2739.82
1780,Larry,Sharp,2689.6499999999996
986,Catherine,Hawkins,2629.9
9676,Theresa,Smith,2599.84
1847,Mary,Smith,2589.87
11901,Mary,Smith,2469.87
4618,Andrea,Smith,2429.8199999999997
10896,Victoria,Smith,2419.78


### 4. Revenue Per Category

Get the revenue generated for each category for the month of 2014 January
* Tables - `orders`, `order_items`, `products` and `categories`
* Data should be sorted in ascending order by `category_id`.
* Output should contain all the fields from `categories` along with the revenue as `category_revenue`.
* Consider only `COMPLETE` and `CLOSED` orders

In [65]:
%%sql 

SELECT c.category_id, c.category_department_id, c.category_name,
    SUM(round(oi.order_item_subtotal::numeric, 2)) AS category_revenue
FROM orders AS o 
    JOIN order_items AS oi
        ON o.order_id = oi.order_item_order_id
        AND to_char(o.order_date::timestamp, 'yyyy-MM') = '2014-01'
        AND o.order_status IN ('COMPLETE', 'CLOSED')
    JOIN products AS p 
        ON oi.order_item_product_id = p.product_id
    JOIN categories AS c 
        ON p.product_category_id = c.category_id
GROUP BY 1, 2, 3
ORDER BY 1 ASC;

Running query in 'postgresql://itversity_retail_user:***@localhost:5432/itversity_retail_db'

33 rows affected.

category_id,category_department_id,category_name,category_revenue
2,2,Soccer,1094.88
3,2,Baseball & Softball,3214.41
4,2,Basketball,1299.98
5,2,Lacrosse,1299.69
6,2,Tennis & Racquet,1124.75
7,2,Hockey,1433.00
9,3,Cardio Equipment,133156.77
10,3,Strength Training,3388.96
11,3,Fitness Accessories,1509.73
12,3,Boxing & MMA,3998.46


### 5. Product Count Per Department

Get the count of products for each department.
* Tables - `departments`, `categories`, `products`
* Data should be sorted in ascending order by `department_id`
* Output should contain all the fields from `departments` and the product count as `product_count`

In [66]:
%%sql 

SELECT * FROM departments limit 0;

Running query in 'postgresql://itversity_retail_user:***@localhost:5432/itversity_retail_db'

department_id,department_name


In [68]:
%%sql 

SELECT d.department_id, d.department_name, COUNT(*) AS product_count
FROM products AS p 
    JOIN categories AS c 
    ON p.product_category_id = c.category_id
    JOIN departments AS d 
    ON c.category_department_id = d.department_id
GROUP BY 1, 2
ORDER BY 1 ASC;

Running query in 'postgresql://itversity_retail_user:***@localhost:5432/itversity_retail_db'

6 rows affected.

department_id,department_name,product_count
2,Fitness,168
3,Footwear,168
4,Apparel,140
5,Golf,120
6,Outdoors,336
7,Fan Shop,149
